code/Controllers/ModelAsController.php

```diff
@@ -37,7 +37,7 @@ class ModelAsController extends Controller implements NestedController {
 	 * @return ContentController
 	 */
 	public static function controller_for(SiteTree $sitetree, $action = null) {
-		$controller = $sitetree->getControllerName();
+		$controller = $sitetree->getFrontendControllerName();
```

Comment:
Suggest that this becomes `getFrontendController()` and includes the Injector call within that method.